### DISCUSSION TITLE
0076 XLS-76d: Min Incoming Amount

### DISCUSSION THREAD

**Discussion by xrpl365** [XRPLF/XRPL-Standards#219] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by tequdev** [comment]
When doing a token swap, meaning Account=Destination Payment, there would be no need to check using the new field.

**Reply by xrpl365** [reply]
You mean where you are sending to yourself, you don't apply the limit?
If so, yes, this would be an easy check to add.
As for its value I would be interested to hear other perspectives.
My own view is that if the overhead of the check is minimal, then it doesn't hurt, it merely adds flexibility.

**Comment by mvadari** [comment]
What about other transactions with a destination that also send funds, like `EscrowCreate`, `PaymentChannelCreate`, and `CheckCreate`?

**Reply by xrpl365** [reply]
I suppose we would have to evaluate the need to support other tx types.
If we were to remove the check for transaction type and simply test any incoming tx with an amount in "xrp" then it would quite possibly reduce the overhead as its one less test, and protect them all. However, that might add complications down the line. If this proposal were to gain traction though, this question would certainly need to be considered more thoroughly. Thanks for the input.

**Reply by KhaledElawadi** [reply]
For sure we should anticipate a need to add other wallet filters and features as different types of tokenised assets are processed on chain.

An interesting idea of yours Chris to expand the remit to allow the user to apply the same set XRP value globally to other tokens. 
That could eliminate spam/dusting from all token issuers and users at the same time. 👍

**Reply by xrpl365** [reply]
I don't think we would apply the min amount to all currencies, just XRP.
I would prefer to find a solution that allowed the user to set this on a token basis, so it makes sense to consider adding to the trust line, but... I've really not given it that much thought cuz I decided it's best to keep it simple and just support XRP first.

Somewhere on the doc I do say that if support for other currencies is deemed worthwhile the spec could be expanded or a second amendment created.

**Reply by KhaledElawadi** [reply]
Agreed, I think it's very wise to apply the minimum threshold filter to XRP first.

However, we should anticipate that spammers might choose to create their own tokens and/or switch to other tokenised assets in the future.

Flexibility could be key in those scenarios.

**Reply by KhaledElawadi** [reply]
Ok, so quick update from some X conversations..
@TheShillVerse loves this idea and has also requested an additional filter which he refers to as "Max Memo Size".

He would like the option of a low pass filter based upon a user adjustable maximum threshold size, so I will make a note of that and add it to my list XRPL Wallet Filters.

**Reply by EnchStyle** [reply]
> I don't think we would apply the min amount to all currencies, just XRP. I would prefer to find a solution that allowed the user to set this on a token basis, so it makes sense to consider adding to the trust line, but... I've really not given it that much thought cuz I decided it's best to keep it simple and just support XRP first.
> 
> Somewhere on the doc I do say that if support for other currencies is deemed worthwhile the spec could be expanded or a second amendment created.

I would like to see it for all tokens. Because I can see how people will just switch from sending XRP to sending tokens. Yes, less reach, but can be even more targeted and efficient. And no extra cost really.

**Comment by xVet** [comment]
Congrats on the first spec 🤗

Would this be easier done on the wallet UI/UX level?

It's a cool spec!

**Reply by xrpl365** [reply]
Don't see how you can prevent an incoming transaction from a wallet?

Wallets could prevent sending but anyone motivated to dust could easily get around that and in fact likely does it with custom code already.

**Reply by xVet** [reply]
> Don't see how you can prevent an incoming transaction from a wallet?
> 
> Wallets could prevent sending but anyone motivated to dust could easily get around that and in fact likely does it with custom code already.

It would still land in your wallet but not bothering you by filtering it within the wallet UI e.g "Show only incoming transactions >1xrp"

I don't think wallets currently allow that, would be a cool addition after reading your proposal 🤗

**Reply by xrpl365** [reply]
Yes that is true, a wallet could be configured not to show them.

**Reply by KhaledElawadi** [reply]
Yes, that could be another optional wallet feature.

However, masking this problem doesn't protect users or the network from attacks.

As we know part of the issue of spam attacks on the network is the fee escalations which affect everyone including those who make lots of legitimate transactions.
A few bad actors over a period of time could cost network users as a whole potentially millions.

**Reply by KhaledElawadi** [reply]
That's not to mention speed and stability.

**Reply by WietseWind** [reply]
That would eliminate the what is in my opinion most important use case: not to let others spam and full up your TX history with crap. In the grand scheme of things, long term, etc. that shouldn't matter because the UI/tools is what should show the users what they need, no matter the raw data on the underlying rails - but spam making spamming less feasible is a pretty good reason for adding this on the core layer imo.

**Reply by KhaledElawadi** [reply]
Thanks for your welcomed input.

I take your point about future UI/tools and account/wallet enhancements. So much of what is possible relies on a good user interface that looks good, customisable, intuitive and user friendly.

I have an upgrade on my list that is so ticklish, in some ways so obvious yet incredibly powerful and was thinking of revealing it a bit later.
Then to my surprise I get a ping and discover someone is actually working on it right now though it's difficult to tell how they're progressing, perhaps slowly?

Needless to say it could revolutionise adoption in so many ways.

These are exciting times in history of the XRPL and all those who can contribute should even if it acts only to stimulate others into forming new creative ideas.

**Reply by xrpl365** [reply]
At a technical level I don't think that it prevents us from having to store the memo on the rejected transaction. However, take away the value in spamming and maybe there is a drop in spam over the long term. 

However this does require us to make sure there are no workarounds that can be used, as Mayukha queried... like using an escrow or other tx.

**Reply by xVet** [reply]
> That would eliminate the what is in my opinion most important use case: not to let others spam and full up your TX history with crap. In the grand scheme of things, long term, etc. that shouldn't matter because the UI/tools is what should show the users what they need, no matter the raw data on the underlying rails - but spam making spamming less feasible is a pretty good reason for adding this on the core layer imo.

That's a good perspective of seeing it. 

I agree with that.

**Reply by mvadari** [reply]
> That would eliminate the what is in my opinion most important use case: not to let others spam and full up your TX history with crap.

It helps on a per-user basis, but it doesn't help with the full-history/infra storage problem - these transactions will still have to be stored, since they'll require a `tec` error.

**Reply by KhaledElawadi** [reply]
Trying to understand the process..

How is the tec error code stored?

Is it as a reference that is associated with a failed transaction and is that reference code stored inside another transaction somehow?

Just trying to work out how and why you would want to store a reference to a transaction that never took place?

**Reply by mvadari** [reply]
Here's an example transaction: https://livenet.xrpl.org/transactions/DBB3C377E14477073E727A23D6826EB3BA6562E1F4F623405B9B7E29C10E8BDB

Transactions with a `tec` error are stored because they change the ledger state - they charge a transaction fee and don't do anything else. So they did take place, they just weren't successful.

**Reply by KhaledElawadi** [reply]
LoL

I think you know what I'm going to say next to that ridiculous state of affairs.

Why are we accepting payments for transactions that never took place?

If I go to a post office and try to pay a bill using my card and there's a small fee for doing so, if during the process the transaction fails I'm not charged the payment fee. 
So why is our ledger taking payments for transactions that never take place?

**Reply by xrpl365** [reply]
> LoL
> 
> 
> 
> I think you know what I'm going to say next to that ridiculous state of affairs.
> 
> 
> 
> Why are we accepting payments for transactions that never took place?
> 
> 
> 
> If I go to a post office and try to pay a bill using my card and there's a small fee for doing so, if during the process the transaction fails I'm not charged the payment fee. 
> 
> So why is our ledger taking payments for transactions that never take place?

We charge a fee because the transaction is using ledger resources, it's a fundamental part of how the ledger protects itself. If you removed that you could do a whole bunch of transactions you know will fail, congest the network and pay no penalty.

It's important to note that the primary purpose of this proposal is not to prevent memo spam, that is merely a positive potential byproduct.

Preventing memo spam is going to require a different type of approach.

**Reply by mvadari** [reply]
You are charged because the network has to do a certain amount of computation before the transaction fails. It prevents spam and DoS attacks - otherwise you could overload all the nodes by making them do a ton of computation for free. 

There are plenty of examples of being charged for failed transactions even in the real world - charging a lawyer for their time even if you lose the case, credit card processing fees even for chargebacks, or various other platform processing fees that are nonrefundable.

**Reply by KhaledElawadi** [reply]
Ok, so I think I understand it more clearly now.

So, there are network resources being used just to attempt a payment and there are network resources used to collect a fee on a payment that never took place.

So, the high pass filter blocks the spammers low cost payment but the protocol still collects a fee.

I will give this some thought though my first suggestion would be that the spammers account must be limited by the amount of the failed transactions he makes per second which have been blocked by our filter.

For example, he attempts to make 1000 transactions in a short amount of time fir one account and the 1st, 2nd, 3rd, *insert amount here* failed attempt the ledger applies a time out similar to a screen lock on a smart phone.

**Reply by mvadari** [reply]
That's not feasible from a technical standpoint. It also doesn't solve the problem, because it doesn't matter if it's one person or a thousand people who send a total of 1000 transactions - it's still going to take the same amount of network resources, and we need to charge for each use. If a person submits one transaction per second or 100 transactions per second, it doesn't matter - they should all be charged.

**Reply by xrpl365** [reply]
> Ok, so I think I understand it more clearly now.
> 
> 
> 
> So, there are network resources being used just to attempt a payment and there are network resources used to collect a fee on a payment that never took place.
> 
> 
> 
> So, the high pass filter blocks the spammers low cost payment but the protocol still collects a fee.
> 
> 
> 
> I will give this some thought though my first suggestion would be that the spammers account must be limited by the amount of the failed transactions he makes per second which have been blocked by our filter.
> 
> 
> 
> For example, he attempts to make 1000 transactions in a short amount of time fir one account and the 1st, 2nd, 3rd, *insert amount here* failed attempt the ledger applies a time out similar to a screen lock on a smart phone.

I love how your mind attempts to problem solve and in a more traditional application (non-DLT) this would be fine but every piece of functionality be it a new cool feature or something aimed at protection must be weighed against the resources required to implement. What I mean by this is that keeping track of the number of rejected transactions over a period of time, comes at a cost, disk I/O, computation and analysis. Whilst ideas seem pretty straightforward on the surface they are complicated significantly by the constraints of the distributed design of DLT. So I'm not saying it's not possible, just that on balance the problem isn't large enough for the resources required and just remember that resources used for this would be resource you can't use for some other process.

I hope this helps.
Now let's keep the comments on the topic of the proposal for the benefit of others reading.

**Reply by KhaledElawadi** [reply]
Ok, well the cool thing I like about this proposal is that the user has total control over their accounts and can adjust each threshold accordingly to suit their particular needs.

**Reply by KhaledElawadi** [reply]
> That's not feasible from a technical standpoint. It also doesn't solve the problem, because it doesn't matter if it's one person or a thousand people who send a total of 1000 transactions - it's still going to take the same amount of network resources, and we need to charge for each use. If a person submits one transaction per second or 100 transactions per second, it doesn't matter - they should all be charged.

Ok, so in another proposal where we discuss the merits of removing infrequent users and legitimate DID verified businesses from paying network fees and I explained how that leaves only a small percentage of users prone to fee escalation in the event of an attack.

Now I'm thinking we could use 76d to trigger an instant fee escalation on an attacker's account which is most likely to be spamming other accounts indiscriminately at the same time instead of suspending it with a time out.

Maybe this could provide the targeted surgical strike that we require instead of the carpet bombing effect currently being deployed where everyone suffers?

**Reply by KhaledElawadi** [reply]
Here's how it works..

Each time an attacker tries to send an ultra low value payment to an account protected by 76d it adds another fee excalation of 100% to that attacker's account.

The more times the attacker's account is blocked by 76d the more escalation is applied.

The beauty of this 76d fee escalation is that it acts like a sentinel waiting for attacks, blocking the attacks and instantly escalates the fee on the attacker's account all at the same time. 👍

**Reply by xrpl365** [reply]
> Here's how it works..
> 
> Each time an attacker tries to send an ultra low value payment to an account protected by 76d it adds another fee excalation ofc100% to thst attackers account.
> 
> The more times the attacker's account is blocked by 76d the more escalation is applied.
> 
> The beauty of this 76d fee escalation is that it acts like a sentinel waiting for an attacks, blocks the attacks and instantly escalates the fee on the attacker's account all at the same time. 👍

If a user engages the feature and they choose to reject low payments, thats fine, its their right to do that.
But not all low payments are spam and it is likely that some genuine payments could get mixed up with the spam.
Implementing what you described would then start to impact a genuine user accounts with escalated fees, just because someone blocked it; the action of blocking is not proof of spam, yet you treat it as such. I cannot see how this can be introduced safely without having unintended consequences.

**Reply by KhaledElawadi** [reply]
They are unsolicited ultra low payments.

Who on the XRPL is legitimately making those types of unwanted payments?

We're talking about many fractions of a penny.

**Reply by xrpl365** [reply]
> They are unsolicited ultra low payments.
> 
> 
> 
> Who on the XRPL is legitimately making those types of unwanted payments?
> 
> 
> 
> We're talking about many fractions of a penny.
> 
> 
> 
> 
> 
> 
> 
> 

Someone who is using the memo for a real use case and is doing a low value payment to achieve it. Or it could be some king of future micro transaction. You can't paint with a broad brush.

**Reply by ximinez** [reply]
> Who on the XRPL is legitimately making those types of unwanted payments?

Scenario 1
1. Scammer sets up an account with a limit of 10,000 XRP.
2. Scammer runs their scam.
3. Victims sending more than the limit get ripped off.
4. Victims sending less than the limit now have their account fees permanently(?) messed up with no way to fix them.

Scenario 2
1. Attacker sets up an account with a limit of 100 XRP.
2. Attacker has funds on an exchange. Exchange has not updated their software to check for minimum limits.
3. Attacker tries to send 50 XRP from their exchange account to their account.
4. Repeat 1000 times.
5. Exchange burns their entire operating account's XRP balance on fees. (10*2^1000 is greater than 10^16, which is about 1/10 of all the XRP in existence.)
6. If exchange manages to make itself whole, it's still stuck paying exorbitant fees to use it's operating account. 

Evil people are creative. You've gotta stay one step ahead of them.

**Reply by ximinez** [reply]
> It helps on a per-user basis, but it doesn't help with the full-history/infra storage problem - these transactions will still have to be stored, since they'll require a `tec` error.

@mvadari The error doesn't necessarily have to be a `tec`, at least when applying against an open ledger. This would prevent the transaction from being written to the ledger, and would prevent it from propagating. If a clever attacker orders transactions such that their transaction _does_ propagate, then they'd pay a fee.

**Reply by KhaledElawadi** [reply]
> > Who on the XRPL is legitimately making those types of unwanted payments?
> 
> Scenario 1
> 
> 1. Scammer sets up an account with a limit of 10,000 XRP.
> 2. Scammer runs their scam.
> 3. Victims sending more than the limit get ripped off.
> 4. Victims sending less than the limit now have their account fees permanently(?) messed up with no way to fix them.
> 
> Scenario 2
> 
> 1. Attacker sets up an account with a limit of 100 XRP.
> 2. Attacker has funds on an exchange. Exchange has not updated their software to check for minimum limits.
> 3. Attacker tries to send 50 XRP from their exchange account to their account.
> 4. Repeat 1000 times.
> 5. Exchange burns their entire operating account's XRP balance on fees. (10*2^1000 is greater than 10^16, which is about 1/10 of all the XRP in existence.)
> 6. If exchange manages to make itself whole, it's still stuck paying exorbitant fees to use it's operating account.
> 
> Evil people are creative. You've gotta stay one step ahead of them.

In scenario 1 doesn't make much sense.

If scammers are wanting muliple payments from their victims then why would they set their min threshold so high?

Anyone that sends payments that are lower than the threshold to an account they are unaware of the threshold gets their asset back and the network fee for their next payment has been ratcheted up one notch on their account and their account only.
Could have a 24 hour reset on that as I don't see why genuine mistakes should be punished I definitely.

In scenario 2 you're simply talking about high frequency payments between two accounts and a reason why the general fee escalation protocol should only apply to individual accounts sending payments and not the entire network as a whole.

**Reply by EnchStyle** [reply]
> Yes, that could be another optional wallet feature.
> 
> However, masking this problem doesn't protect users or the network from attacks.
> 
> As we know part of the issue of spam attacks on the network is the fee escalations which affect everyone including those who make lots of legitimate transactions. A few bad actors over a period of time could cost network users as a whole potentially millions.

This also protect companies during taxation processes, where you have to use like Koinly and they hit you with a limit, because instead of 5000 transactions you now how 20000, because for each outgoing you get 2-3 incoming.

**Comment by LadyK-0** [comment]
I've actually talked to some people about the xrp dusting attacks and they literally thought it carried the same security risks that it would on the Ethereum network, but we don't have smart contracts, so on the xrpl you just received an asset(unwanted.) I do wonder the percentage of people who think that. However I think this amendment would have a positive affect on the perception of preventing dusting attacks. But also create other negative feedback. In an environment where there is open permissionless participation you just can't stop people from doing negative things. So everyone sets a minimum xrp acceptance, what if they start sending dusting attacks on other token trustlines, not as many holders but still possible even with this amendment. This is what the fee escalation is suppose to prevent right, the spamming of the ledger. Overall I don't think this amendment would hurt anything. Unless creating an extra check on accounts and transactions would slow it. Just some thoughts.

**Reply by KhaledElawadi** [reply]
Thanks Lady K.

You're right there isn't currently anything preventing dusters/spammers from using other low value tokens to mess up people's accounts with unsolicited payments. We can modify this proposal to include all assets and all transfer/payment methods.

The fee escalation protocol is slightly different thing that we can also upgrade namely because it's a global punishment that hits everyone. 
For an attacker/saboteur that's classed as a win. I've made a separate proposal on X on how we can put a stop to that too.
You can find it at this address:

https://x.com/KhaledElawadi/status/1827294767413149932?t=RIh-e4WWiI-zKzmG0vqAxg&s=19

**Reply by xrpl365** [reply]
> I've actually talked to some people about the xrp dusting attacks and they literally thought it carried the same security risks that it would on the Ethereum network, but we don't have smart contracts, so on the xrpl you just received an asset(unwanted.) I do wonder the percentage of people who think that. However I think this amendment would have a positive affect on the perception of preventing dusting attacks. But also create other negative feedback. In an environment where there is open permissionless participation you just can't stop people from doing negative things. So everyone sets a minimum xrp acceptance, what if they start sending dusting attacks on other token trustlines, not as many holders but still possible even with this amendment. This is what the fee escalation is suppose to prevent right, the spamming of the ledger. Overall I don't think this amendment would hurt anything. Unless creating an extra check on accounts and transactions would slow it. Just some thoughts.

Protecting other currencies or even or types of tx is likely possible too but I'd like to see the kind of take up a small amendment like this had before adding more functionality. You're right about the effect on performance, and this would certainly need to be considered, benchmarked and tested but on first review I doubt the overhead would be significant.

Thanks for contributing your feedback.

**Comment by Mwni** [comment]
If the big wallets like Xaman backed this and encouraged their users to set such a minimum amount, it could effectively eliminate the spam problem.

**Reply by KhaledElawadi** [reply]
Agreed.

In a way we would all be securing the network against targeted attacks by using the filter to block them.

I like the fact the threshold can also be varied by the user should a well funded attack come in.

**Reply by KhaledElawadi** [reply]
In light of the new and exciting XLS - 86d Firewall proposal I've agreed with Chris that this feature could easily be bundled into a suite of Firewall Protection.

In addition I suggest that a whitelist option is added to it because some users may wish to receive low value payments/fees streamed to them on a regular basis from certain trusted accounts whilst simultaneously blocking unsolicited spam.

**Reply by KhaledElawadi** [reply]
Ok, so I've noticed a type of scam that's become very popular recently that's called "Address Poisoning".

This is where the scammers send ultra low micro payments to accounts that have very similar addresses with usually the first and/or last 4-6 character/numbers. Sometimes done immediately after making transfers to or from centralised exchanges and/or user's alternative accounts.

The idea is to trick users who don't pay close enough attention into copy and pasting the scammers account address instead of the intend one.

This amendment with inclusive updates for other issued tokens such RWAs will protect XRPL user accounts from being poisoned.

**Comment by shortthefomo** [comment]
Wonderful idea LGTM

+1 for "no limit" on **Maximum limit** as this also opens up other use cases for businesses.

**Comment by xrpl365** [comment]
This discussion has been closed and moved to PR: https://github.com/XRPLF/XRPL-Standards/pull/355

For reference the functionality provided by this amendment will now move forward as part of Firewall - PR:  https://github.com/XRPLF/XRPL-Standards/pull/363
